(DOCSP-27219): (Swift SDK) Add a Write to a Synced Realm page (#2484)
mongodb/docs-realm: source/sdk/swift/sync/write-to-synced-realm.txt

in behavior to keep in mind as you develop your application.

When you write to a synced realm, your write operations must match *both*
of the following:

- **The sync subscription query.**
    - If your write operation doesn't match the query in the subscription,
      the write reverts with a non-fatal compensating write error (ErrorCompensatingWrite).
- **The Device Sync permissions** in your App Services App.
    - If your try to write data that doesn't match the Device Sync permissions expression,
      the write reverts with a non-fatal permission denied error.
      In the client, this shows as an (ErrorCompensatingWrite). On the server,
      you can see more details about how the write was denied was by a 
      write filter in the role.
    - To learn more about configuring Device Sync permissions for your app,
      refer to :ref:`sync-rules` and the :ref:`flexible-sync-permissions-guide`
      in the App Services documentation.

To learn more about permission denied errors, compensating write errors
and other Device Sync error types, refer to :ref:`sync-errors` in the App Services documentation.

Determining What Data Syncs

> ---------------------------

Reviewer: I think it'd be clearer to call this smthn like 'Examples on this Page' to make clear that this is just an example config and explanation of it.

Author: I renamed this away from "Examples on this Page" because I wanted it to be more about you as a developer understanding what data syncs, and why. And to emphasize that it's because of these two components - the Device Sync Configuration w/permissions and the client model/subscription query. It's really the intersection of these things that's important.

So I think the fact that these are examples that support the code on the page is secondary; I think the important aspect is that the intersection of these two things determines what data syncs.

That said, I kept the intro paragraph from when this was more focused on the "Examples on this Page" title so I could tweak that a bit. I'll rework this a bit and then I'd love to hear whether that makes sense or if you still think it's unclear or otherwise bad naming.

Reviewer: great changes! 

i think i previously had a slightly off mental model, where i was incorrectly thinking that only writes that don't match the query subscription are compensating writes. and then permission denied reverts are a separate but similar thing. 

this clarifies things for me (and hopefully our users too 😄). will soon update the flutter page w this in mind.

The examples on this page use an Atlas App Services App with the following 
Device Sync configuration and a client app with the following Realm SDK 
data model and subscriptions.

Device Sync Configuration
~~~~~~~~~~~~~~~~~~~~~~~~

Device Sync is configured with the following queryable fields:

> - ``_id`` (this is always a queryable field in every Device Sync configuration)

Reviewer: ```suggestion
- ``_id`` (this is a queryable field in every Device Sync configuration)
```

soft nit: felt a bit redundant to have 'always' and 'every'. could also **bold** 'every' for more emphasis

- ``ownerId``
- ``complexity``

Device Sync has permissions configured to let users read and write only their own
data:

.. code-block:: json

   {
     "rules": {},
     "defaultRoles": [
       {
         "name": "owner-read-write",
         "applyWhen": {},
         "read": {
           "ownerId": "%%user.id"
         },
         "write": {
           "ownerId": "%%user.id"
         }
       }
     ]
   }

Client Data Model and Configuration

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer: ```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

The examples on this page use the following object model:

.. literalinclude:: /examples/generated/code/start/SyncedRealmCRUD.snippet.flexible-sync-crud-model.swift
   :language: swift

Using that object model, the synced realm configuration syncs objects that 
match the subscription query where the ``complexity`` property's value 
is less than or equal to ``4``:

.. literalinclude:: /examples/generated/code/start/SyncedRealmCRUD.snippet.flexible-sync-subscription-setup.swift
   :language: swift

What Data Syncs?
~~~~~~~~~~~~~~~~

The subscription query combined with the Device Sync permissions mean
that the synced realm only syncs objects where:

- The ``ownerId`` matches the ``user.id`` of the logged-in user (from the permissions)
- The ``complexity`` property's value is less than or equal to ``4`` (from the subscription query)

Any object in the Atlas collection where the ``ownerId`` does not match 
the ``user.id`` of the logged-in user, or the ``complexity`` property's 
value is greater than ``4``, cannot sync to this realm.

Write to a Synced Realm

> -----------------------

Reviewer: i think the subsections here could use a bit of refactoring w/o necessary making copy changes. 

the core of the confusion for me while reading this came from two things:
- the long description of the writes failing process before discussion of successful writes or the ways that the writes fail below.
- lack of conceptual description of 'reverted write', which can be caused by the permission denied or writing outside the query subscription.

for me, more logical structure here would be (pardon the Markdown but i think it makes the point more clearly about the headers):

````md
## Write to a Synced Realm

When you write to a synced realm, these are the most common cases:

- **Successful writes**: The written object matches both the query subscription
  and the user's permissions. The object writes successfully to the realm,
  and syncs successfully to the App Services backend and other devices.
- **Reverted writes**: If you perform a write that does not match
  the subscription query, or where the  user does not have sufficient permissions to 
  perform the write, Realm reverts the illegal write.

### Successful Writes

*keep content in this section the same*

### Reverted Writes

If you perform a write that does not match the subscription query, or where
the user does not have sufficient permissions to perform the write, Realm 
performs a compensating write. This is a non-fatal operation that 
essentially reverts an illegal write. The two most common reasons for reverted writes are the following: 

- **Writes that don't match the query subscription**: The written object matches 
  the user's permissions, but does not match the query subscription. The
  object writes to the realm, but when the realm syncs to the backend, it
  reverts the write in a non-fatal error operation we call a compensating 
  write.
- **Writes that don't match permissions**: The written object matches the query
  subscription, but does not match the user's permissions. The
  object writes to the realm, but when the realm syncs to the backend, it
  reverts the write in a non-fatal error operation we call a compensating 
  write.

In more detail, when you write data that is outside the bounds of a query
subscription or does not match the user's permissions, the following occurs:

#. Because the client realm has no concept of "illegal" writes,
   the write initially succeeds until realm resolves the changeset
   with the App Services backend.
#. Upon sync, the server applies the rules and permissions.
   The server determines that the user does not have authorization to perform the write.
#. The server sends a revert operation, called a "compensating write", back to the client.
#. The client's realm reverts the illegal write operation.

Any client-side writes to a given object between an illegal write to that object
and the corresponding compensating write will be lost.

In practice, this may "look like" an object being written to the realm, and
then disappearing after the server sends the compensating write back to 
the client.

#### Writes that Don't Match the Query Subscription

*keep content in this section the same*

#### Writes That Don't Match Permissions

*keep content in this section the same*

````

Author: I see what you mean here. I've taken your suggestion and restructured the page a bit. I didn't go with exactly what you suggested here, but it's close - just some minor differences. Curious what you think now.

Reviewer: great changes! 

i think i previously had a slightly off mental model, where i was incorrectly thinking that only writes that don't match the query subscription are compensating writes. and then permission denied reverts are a separate but similar thing. 

this clarifies things for me (and hopefully our users too 😄). will soon update the flutter page w this in mind.

  object writes to the realm, but when the realm syncs to the backend, it
  reverts the write in a non-fatal error operation we call a compensating 
  write.

If you perform a write that does not match the subscription query, or where
the user does not have sufficient permissions to perform the write, Realm 
performs a compensating write. This is a non-fatal operation that 
essentially reverts an illegal write.

In more detail, when you write data that is outside the bounds of a query
subscription or does not match the user's permissions, the following occurs:

#. Because the client realm has no concept of "illegal" writes,
   the write initially succeeds until realm resolves the changeset
   with the App Services backend.
#. Upon sync, the server applies the rules and permissions.
   The server determines that the user does not have authorization to perform the write.
#. The server sends a revert operation, called a "compensating write", back to the client.
#. The client's realm reverts the illegal write operation.

Any client-side writes to a given object between an illegal write to that object
and the corresponding compensating write will be lost.

> In practice, this may "look like" an object being written to the realm, and

Reviewer: why 'look like' in quotes? isn't this literally what's happening?

Author: I went with quotes for a couple of reasons which in retrospect don't really seem to matter so much. Removing the quotes.

then disappearing after the server sends the compensating write back to 
the client.

.. _swift-successful-writes:

Successful Writes
~~~~~~~~~~~~~~~~~

When the write matches both the :ref:`Device Sync permissions <sync-rules>` 
and the :ref:`Flexible Sync subscription query 
<swift-manage-flexible-sync-subscriptions>` in the client, the Realm Swift SDK 
can successfully write the object to the synced realm. This object syncs 
with the App Services backend when the device has a network connection.

.. literalinclude:: /examples/generated/code/start/SyncedRealmCRUD.snippet.successful-write.swift
   :language: swift

.. _swift-compensating-writes:

Writes that Don't Match the Query Subscription
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can only write objects to a Flexible Sync realm if they match the 
subscription query. If you perform a write that does not match the 
subscription query, Realm initially writes the object, but then performs 

> a compensating write. This is a non-fatal operation that essentially 
> reverts an illegal write that does not match the subscription query.

Reviewer: ```suggestion
a compensating write. This is a non-fatal operation that 
reverts an illegal write that does not match the subscription query.
```

> In practice, this may look like the write succeeding, but then the object
> "disappearing" when Realm performs the compensating write.

Reviewer: ```suggestion
In practice, this may look like the write succeeding, but then the object
"disappears" when Realm performs the compensating write.
```

Reviewer: also i liked your phrasing on [lines 217-218](https://github.com/mongodb/docs-realm/pull/2484/files#diff-e2d8c506c5da4619322a5fb7fec7dc60573f5642ba09573913f868f317dabb1fR217-R218) about App Services initiating the the compensating write. i think it clearly conveyed the idea that the App Services sync server triggers the compensating write. could you add that here as well?

   :language: swift

Client Error
````````````

The error message in the client-side logs in this scenario is: 

.. code-block:: console

   Sync: Connection[1]: Session[1]: Received: ERROR "Client attempted a 
   write that is outside of permissions or query filters; it has been 
   reverted" (error_code=231, try_again=true, error_action=Warning)

App Services Error
``````````````````

The error message in the App Services logs in this scenario is:

.. code-block:: console

   "FlexibleSync_Item": {
       "63bdfc40f16be7b1e8c7e4b7": "write to \"63bdfc40f16be7b1e8c7e4b7\" 
       in table \"FlexibleSync_Item\" not allowed; object is outside of 
       the current query view"
   }

.. _swift-writes-outside-permissions:

Writes That Don't Match Permissions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Attempting to write to the client can also trigger a compensating write 
error when the object does not match the user's write permissions. 

On the client, this type of write behaves the same as a write that doesn't 
match the query subscription. In practice, this may look like the write 

> succeeding, but then the object "disappearing" when Realm syncs with the 

Reviewer: ```suggestion
succeeding, but then the object "disappears" when Realm syncs with the 
```

App Services backend and performs the compensating write.

Code Example
````````````

Given the permissions in the Device Sync Configuration detailed above, 
attempting to write an object where the ``ownerId`` property does not match
the ``user.id`` of the logged-in user is not a legal write:

.. literalinclude:: /examples/generated/code/start/SyncedRealmCRUD.snippet.write-outside-permissions.swift
   :language: swift

Client Error
````````````

The client error in this scenario is the same as when you attempt to write
an object that is outside the query filter:

.. code-block:: console

   Sync: Connection[1]: Session[1]: Received: ERROR "Client attempted a 
   write that is outside of permissions or query filters; it has been 
   reverted" (error_code=231, try_again=true, error_action=Warning)

App Services Error
``````````````````

> The error message in the App Services logs provides some additional 
> information to help you determine that it is a permissions issue - in this 
> case the object does not match the user's role - and not a query subscription 
> issue:

Reviewer: ```suggestion
The error message in the App Services logs provides some additional 
information to help you determine that it is a permissions issue,
and not a query subscription issue. In this example, the error message shows
that the the object does not match the user's role:
```

nit: break up sentence for readability.

Overview
--------

When writing data to a synced realm using Flexible Sync, you can use the 
same APIs as writing to a local realm. However, there are some differences
in behavior to keep in mind as you develop your application.

When you write to a synced realm, your write operations must match *both*
of the following:

- **The sync subscription query.**
    - If your write operation doesn't match the query in the subscription,
      the write reverts with a non-fatal compensating write error (ErrorCompensatingWrite).
- **The Device Sync permissions** in your App Services App.
    - If your try to write data that doesn't match the Device Sync permissions expression,
      the write reverts with a non-fatal permission denied error.
      In the client, this shows as an (ErrorCompensatingWrite). On the server,
      you can see more details about how the write was denied was by a 
      write filter in the role.
    - To learn more about configuring Device Sync permissions for your app,
      refer to :ref:`sync-rules` and the :ref:`flexible-sync-permissions-guide`
      in the App Services documentation.

To learn more about permission denied errors, compensating write errors

> and other Device Sync error types, refer to :ref:`sync-errors` in the App Services documentation.

Author: Hmm, good point that it doesn't fit well here anymore - I actually think it fits better in the "Compensating Writes" section (ln 164 ish) so moving it directly into there.

Compensating Writes
~~~~~~~~~~~~~~~~~~~

In some cases, a write that initially appears to succeed is actually an 
illegal write. In these cases, the object writes to the database, but when 
the database syncs to the backend, Realm reverts the write in a non-fatal 
error operation called a compensating write. Compensating writes can occur
when:

- **Writes that don't match the query subscription**: The written object matches 
  the user's permissions, but does not match the query subscription. 
- **Writes that don't match permissions**: The written object matches the query
  subscription, but does not match the user's permissions.

In more detail, when you write data that is outside the bounds of a query
subscription or does not match the user's permissions, the following occurs:

#. Because the client realm has no concept of "illegal" writes,
   the write initially succeeds until realm resolves the changeset
   with the App Services backend.
#. Upon sync, the server applies the rules and permissions.
   The server determines that the user does not have authorization to perform the write.
#. The server sends a revert operation, called a "compensating write", back to the client.

> #. The client's realm reverts the illegal write operation.

Reviewer: i think it could be helpful if in this sentence it makes clear that "reverting illegal write" basically means "delete the illegal object from the realm". so perhaps rephrase sentence to smthn like: 

> The client's realm reverts the illegal write operation by deleting the illegal object from the realm.

Code Example
++++++++++++

Given the configuration for the Flexible Sync realm above, attempting to 
write this object does not match the query subscription:

.. literalinclude:: /examples/generated/code/start/SyncedRealmCRUD.snippet.write-outside-flexible-sync-query.swift
   :language: swift

Client Error
++++++++++++

The error message in the client-side logs in this scenario is: 

.. code-block:: console

   Sync: Connection[1]: Session[1]: Received: ERROR "Client attempted a 
   write that is outside of permissions or query filters; it has been 
   reverted" (error_code=231, try_again=true, error_action=Warning)

App Services Error
++++++++++++++++++

The error message in the App Services logs in this scenario is:

.. code-block:: console

   "FlexibleSync_Item": {
       "63bdfc40f16be7b1e8c7e4b7": "write to \"63bdfc40f16be7b1e8c7e4b7\" 
       in table \"FlexibleSync_Item\" not allowed; object is outside of 
       the current query view"
   }

.. _swift-writes-outside-permissions:

Writes That Don't Match Permissions
```````````````````````````````````

Attempting to write to the client can also trigger a compensating write 

> error when the object does not match the user's write permissions. 

Reviewer: ```suggestion
error when the object does not match the user's server-side write permissions. 
```

nit: clarity where the permissions are
